java/java.hints/src/org/netbeans/modules/java/hints/bugs/Unused.java

```diff
@@ -52,7 +52,12 @@ public class Unused {
     @BooleanOption(displayName="#LBL_UnusedPackagePrivate", tooltip="#TP_UnusedPackagePrivate", defaultValue=DETECT_UNUSED_PACKAGE_PRIVATE_DEFAULT)
     public static final String DETECT_UNUSED_PACKAGE_PRIVATE = "detect.unused.package.private";
 
-    @TriggerTreeKind(Kind.COMPILATION_UNIT)
+    @TriggerTreeKind({
+        //class-like kinds:
+        Kind.ANNOTATION_TYPE, Kind.CLASS, Kind.ENUM, Kind.INTERFACE, Kind.RECORD,
+        Kind.VARIABLE,
+        Kind.METHOD
+    })
     public static List<ErrorDescription> unused(HintContext ctx) {
```

Comment:
@Achal1607 question: why was this change made?

I ran some benchmarks which were based on generated files I used for https://github.com/apache/netbeans/pull/4142

very large files (10K fields and getters) with many positive unused hits performed slightly worse:
PR:
```
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=69114.52ms, invocations=20003, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unused
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: hint processing complete
```
master:
```
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=62578.07ms, invocations=1, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unused
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: hint processing complete
```

the more concerning bit is this though. a 3k fields file which are _not_ unused:

PR:
```
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=565.49ms, invocations=6003, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unused
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: hint processing complete
```
master:
```
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.11ms, invocations=1, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unused
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: hint processing complete
```

Reply:
This change is made so that it can honour `@SuppressWarnings("unused")` otherwise currently suppress warning annotation for unused variables is not working. For more info please refer to https://github.com/oracle/javavscode/issues/96

Reply:
ah got it, the annotation has to anchor on the trigger. I didn't expect that this would matter - makes sense though.

A bit of a shame that this makes the not-unused case so much slower.

Reply:
I am also a bit of surprised by the performance degradation in the not-unused case.

Reply:
the method is going to be called a few k times on large files, compared to one time before the change. But it seems we are lucky here and a simple fast-path is able to get close to the old state:

I added this after `findUnused(...)` which is already caching results internally
```java
        if (unused.isEmpty()) {
            return null;
        }
```
`null` will lead to the fast path in `CodeHintProviderImpl#createErrors`. 

and we get:
```
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.83ms, invocations=6003, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unused
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: hint processing complete
```

At some point we should take a look at this method and check what is causing the slowdown there.

Reply:
it turns out that reading values from Preferences in hot code causes performance problems due to the need of a read lock. This is the reason why the fast path made things so much faster in the not-unused case.

here a commit with more changes: https://github.com/mbien/netbeans/commit/eb1dc47da1f17adcf88fc03af4bf4369c83ed5f1

feel free to squash with your PR